File: ios/NativeSigner/Components/Buttons.swift
//
//  Buttons.swift
//  NativeSigner
//
//  Created by Sveta Goldstein on 13.12.2021.
//

import SwiftUI

struct BigButtonStyle: ButtonStyle {
    var bgColor: Color
    var fgColor: Color

    func makeBody(configuration: Self.Configuration) -> some View {
        configuration.label
            .padding(12)
            .background(bgColor)
            .cornerRadius(8)
            .foregroundColor(fgColor)
    }
}

struct BigButton: View {
    var text: String
    var isShaded: Bool = false
    var isCrypto: Bool = false
    var isDangerous: Bool = false
    var action: () -> Void
    var isDisabled: Bool = false

    var body: some View {
        let accentColor = isCrypto ? Color("Crypto400") : Color("Action400")
        let bgColor = isDisabled
            ? Color("Bg200")
            : isShaded ? Color("Bg300") : accentColor
        let fgColor = isDisabled
            ? Color("Text300")
            : isShaded
                ? isDangerous ? Color("SignalDanger") : accentColor
                : Color("Action600")

        Button(action: action) {
            HStack {
                Spacer()
                Text(text).font(FBase(style: .button))
                Spacer()
            }
        }
        .buttonStyle(BigButtonStyle(bgColor: bgColor, fgColor: fgColor))
        .disabled(isDisabled)
    }
}

File: android/app/src/main/java/io/parity/signer/components/Buttons.kt
package io.parity.signer.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Surface
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import io.parity.signer.ui.theme.*

/**
 * Typical huge button that just wants to be pushed
 */
@Composable
fun BigButton(
	text: String,
	isShaded: Boolean = false,
	isCrypto: Boolean = false,
	isDangerous: Boolean = false,
	action: () -> Unit,
	isDisabled: Boolean = false
) {
	val accentColor =
		if (isCrypto) MaterialTheme.colors.Crypto400 else MaterialTheme.colors.Action400
	val bgColor =
		if (isDisabled) MaterialTheme.colors.Bg200 else if (isShaded) MaterialTheme.colors.Bg300 else accentColor
	val fgColor =
		if (isDisabled) MaterialTheme.colors.Text300 else if (isShaded) if (isDangerous) MaterialTheme.colors.SignalDanger else accentColor else MaterialTheme.colors.Action600

	Surface(
		color = bgColor,
		shape = MaterialTheme.shapes.large,
		/*border = BorderStroke(
			1.dp,
			if (isShaded) fgColor else bgColor
		),*/
		modifier = Modifier
			.clickable(onClick = action, enabled = !isDisabled)
			.padding(vertical = 8.dp)
			.height(44.dp)
	) {
		Row(
			horizontalArrangement = Arrangement.Center,
			verticalAlignment = Alignment.CenterVertically,
			modifier = Modifier.fillMaxWidth(1f)
		) {
			Text(text, style = MaterialTheme.typography.button, color = fgColor)
		}
	}
}

/**
 * Suggest buttons for seed phrase recovery screen
 */
@Composable
fun SeedPhraseButton(word: String, select: () -> Unit) {
	Surface(
		shape = MaterialTheme.shapes.small,
		color = MaterialTheme.colors.Crypto100,
		modifier = Modifier.clickable(onClick = select)
	) {
		Text(
			word,
			style = CryptoTypography.body2,
			color = MaterialTheme.colors.Crypto400,
			modifier = Modifier.padding(horizontal = 12.dp, vertical = 4.dp)
		)
	}
}

/**
 * Small buttons for multiselect screen
 */
@Composable
fun SmallButton(
	text: String,
	isDisabled: Boolean = false,
	action: () -> Unit
) {
	Surface(
		shape = MaterialTheme.shapes.small,
		border = BorderStroke(1.dp, if (isDisabled) MaterialTheme.colors.Text300 else MaterialTheme.colors.Action400),
		color = Color.Transparent,
		modifier = Modifier.clickable(onClick = action, enabled = !isDisabled)
	) {
		Text(
			text,
			style = MaterialTheme.typography.caption,
			color = if (isDisabled) MaterialTheme.colors.Text300 else MaterialTheme.colors.Action400,
			modifier = Modifier.padding(4.dp)
		)
	}
}
